### Check VOMS data: A-30, vehicle inventory, RR-20 data  

This notebook shows the development of the functions that are used in the executable file, `voms_inventory_check.py`

#### Step 1: Compare the A-30 vehicle list to the inventory list in BlackCat
We are replicating the validation steps that the Grants team currently does, in the order that they currently do it, to ensure automating this process is successful. After interviews we know that early on they compare the A-30 vehicle list to the inventory list in BlackCat, and also to the RR-20 reported VOMS vehicles on the RR-20 form.  
  
These are the prototyped checks in this notebook:  
* Check 1. They compare each vehicle's VIN in both places to ensure that all vehicles in the lists are the same.
* Check 2. They make sure that the total "active" vehicles in the inventory are equal to or less than what is reported on the A-30.
* Check 3. They make sure that the total VOMS vehicles reported on the RR-20 is equal to or fewer than the total vehicles on the A-30.
  
The A-30 form is auto-filled with inventory data as it is, *at that time*. It is **not** updated later if the inventory data changes.

---
### TO DO: 
* think through how data will appear (i.e. all orgs in files together, or 1 file per org? Pull in Sharepoint Excel files, or do Sharepoint -> GCS -> python script?), rewrite inputs to functions.
* think through where to output excel files

---


*For now* while deciding on the above, we import excel files of data from i) the current vehicle inventory, ii) 2022 A-30 & iii) RR-20 forms. 
* The vehicle inventory was downloaded from the BlackCat web interface. 
* The A-30 was downloaded from the BlackCat web interface. 
* The RR-20 was downloaded from the BlackCat web interface. ~emailed to us from Black Cat for 2022.~
  
We are assuming that any future API or larger Excel file will have this data in the **exact same format**.  
  
A few data points must match across forms, so we combine those parts of the RR-20 with the accompanying parts of the A-30 into **one data table** upon which to run checks on (columns can be checked against each other).  
  
The rest of the RR-20 and the A-30 will be kept separate to prevent exploding rows by their multiples.


In [1]:
import pandas as pd
import numpy as np 

In [2]:
## Dataset 1. Import the inventory of revenue vehicles from Black Cat.
inventory = pd.read_excel("data/RevenueVehicles_9_2_2023.xlsx", 
                    index_col=None) 

# print(inventory.shape)
# inventory.head(3)

In [3]:
rev_veh_inventory = inventory[['Organization', 'VIN', 'LicensePlate', 'AgencyId', 'Category', 'Status', \
                               'NTD ID', 'Vehicle Type']]

print(rev_veh_inventory.shape)
rev_veh_inventory.head()

(3824, 8)


Organization                VIN LicensePlate AgencyId  \
0             3.00  1FD3E35LX8DB57263          NaN      NaN   
1             3.00  1FD3E35LX8DB57264          NaN      NaN   
2  Able Industries  1FDEE3FL7FDA08586      43980V1      NaN   
3   AbleLight Inc.  1FMNE1BW1BDA36060      6PCV410   A36060   
4         Alchemia  523MF1A60CM101085          NaN      NaN   

                           Category  Status NTD ID      Vehicle Type  
0          Light-Duty Mid-Sized Bus  Active    NaN  CU - Cutaway Bus  
1          Light-Duty Mid-Sized Bus  Active    NaN  CU - Cutaway Bus  
2              Heavy-Duty Small Bus  Active    NaN  CU - Cutaway Bus  
3  Light-Duty Vans, Sedans or Buses  Active    NaN      MV - Minivan  
4  Light-Duty Vans, Sedans or Buses  Active    NaN          BU - Bus

In [4]:
### Dataset 2. Import the a30 and compare.
a30 = pd.read_excel("data/A_30_Revenue_Vehicle_Report_9_1_2023.xlsx", index_col=None)
a30.head(20)

Organization  Group Plan                 VIN  \
0   Alpine County Community Development         NaN   1FDES8PG1LKB22727   
1   Alpine County Community Development         NaN   1GB3G2BL5E1199290   
2                        Amador Transit         NaN   2C7WDGBG5GR372437   
3                        Amador Transit         NaN   1FDFE45S29DA54849   
4                        Amador Transit         NaN   1FDFE4FS4KDC69024   
5                        Amador Transit         NaN   1FDVU4XG6KKB16858   
6                        Amador Transit         NaN   1GB6G5BL0D1113431   
7                        Amador Transit         NaN   1GB6G5BL0D1114456   
8                        Amador Transit         NaN   1GB6G5BL2D1113415   
9                        Amador Transit         NaN   1GB6G5BL3D1112385   
10                       Amador Transit         NaN   1GB6G5BL8D1114035   
11                       Amador Transit         NaN   2C7WDGBG6FR614294   
12                       Amador Transit         NaN   2C7WDGBG9FR642882   
13                       Amador Transit         NaN   4UZADRDU4ECFR9869   
14                       Amador Transit         NaN   4UZADRDU8ECFN9407   
15                       Amador Transit         NaN   4UZADRFD1JCJW7194   
16                       Amador Transit         NaN   4UZADRFD3JCJW7195   
17                       Amador Transit         NaN   4UZADRFD8JCJW7192   
18                       Amador Transit         NaN   4UZADRFDXJCJW7193   
19            Calaveras Transit Agency          NaN   1FDFE4FS2GDC50284   

   RVI ID ADA Accessible Vehicles (0/No 1/Yes) Vehicle Type Code  \
0     NaN                                  Yes  CU - Cutaway Bus   
1     NaN                                  Yes  CU - Cutaway Bus   
2     NaN                                  Yes      MV - Minivan   
3     NaN                                  Yes  CU - Cutaway Bus   
4     NaN                                  Yes  CU - Cutaway Bus   
5     NaN                                  Yes          VN - Van   
6     NaN                                  Yes  CU - Cutaway Bus   
7     NaN                                  Yes  CU - Cutaway Bus   
8     NaN                                  Yes  CU - Cutaway Bus   
9     NaN                                  Yes  CU - Cutaway Bus   
10    NaN                                  Yes  CU - Cutaway Bus   
11    NaN                                  Yes      MV - Minivan   
12    NaN                                  Yes      MV - Minivan   
13    NaN                                  Yes  CU - Cutaway Bus   
14    NaN                                  Yes  CU - Cutaway Bus   
15    NaN                                  Yes  CU - Cutaway Bus   
16    NaN                                  Yes  CU - Cutaway Bus   
17    NaN                                  Yes  CU - Cutaway Bus   
18    NaN                                  Yes  CU - Cutaway Bus   
19    NaN                                  Yes  CU - Cutaway Bus   

                                       Funding Source  \
0                                           State Aid   
1       Section 5311 – Statewide Rural Public Transit   
2   Section 5310 – Enhanced Mobility of Seniors & ...   
3                  Section 5317 – New Freedom Program   
4                                                 NaN   
5   Other Federal Funds: Section 5337- State of Go...   
6   Section 5310 – Enhanced Mobility of Seniors & ...   
7   Section 5310 – Enhanced Mobility of Seniors & ...   
8   Section 5310 – Enhanced Mobility of Seniors & ...   
9   Section 5310 – Enhanced Mobility of Seniors & ...   
10  Section 5310 – Enhanced Mobility of Seniors & ...   
11                                                NaN   
12  Section 5310 – Enhanced Mobility of Seniors & ...   
13                                                NaN   
14                                                NaN   
15                                                NaN   
16                                                NaN

In [5]:
a30.shape

(742, 15)

### Check 1. Compare each vehicle's VIN in both places to ensure that all vehicles in the lists are the same.

*Output option 1: FULL CHECKLIST WITH ALL VINS*  
Plan is to show the grants and 5311 liaision team these output options to get their feedback.

In [10]:
## Inventory: For each agency and year, get all VINS that are active. 
## A30: For each agency and year, get all VINS 
## both:  Compare VIN list with inventory VIN active list
## if matches, spit out totals (e.g., 17 active) and in description col say "all match"
## if does not match, spit out: from inventory and A30, VIN, status. 

inv_agencies = rev_veh_inventory['Organization'].unique()
a30_agencies = a30['Organization'].unique()

output = []
for agency in a30_agencies:
    if len(a30[a30['Organization']==agency]) > 0:
        vins_a30 = a30[a30['Organization']==agency]['VIN'].unique() #will list only active ins
        vins_inventory = rev_veh_inventory[(rev_veh_inventory['Organization']==agency) \
                 & (rev_veh_inventory['Status']=='Active')]['VIN'].unique()
        
        for v in vins_a30: #check whether each VIN exists in inventory
            if v in vins_inventory:
                description = "Matched an Active vehicle in vehicle inventory."
                result = "Y"
            else:
                description = f"{v} not an active vehicle in the inventory. Investigate."
                result = "N"

            output_line = {"Organization": agency,
                           "VIN": v,
                           "check_status": result,
                          "Description": description}
            output.append(output_line)
    
    full_vin_checklist = pd.DataFrame(output).sort_values(by="Organization")


In [11]:
print(full_vin_checklist.shape)
full_vin_checklist

(741, 4)


Organization                VIN  \
0              Alpine County Community Development  1FDES8PG1LKB22727   
1              Alpine County Community Development  1GB3G2BL5E1199290   
18                                  Amador Transit  4UZADRFDXJCJW7193   
17                                  Amador Transit  4UZADRFD8JCJW7192   
16                                  Amador Transit  4UZADRFD3JCJW7195   
..                                             ...                ...   
734  Yosemite Area Regional Transportation System    1M8PDMBA5CP01268   
733  Yosemite Area Regional Transportation System    1M8PDMBA3CP05986   
732  Yosemite Area Regional Transportation System    1M8PDMBA3CP01268   
735  Yosemite Area Regional Transportation System    1M8PDMBA5CP05986   
740  Yosemite Area Regional Transportation System    1M8PDMEA1AP05940   

    check_status                                        Description  
0              Y    Matched an Active vehicle in vehicle inventory.  
1              N  1GB3G2BL5E1199290 not an active vehicle in the...  
18             Y    Matched an Active vehicle in vehicle inventory.  
17             Y    Matched an Active vehicle in vehicle inventory.  
16             Y    Matched an Active vehicle in vehicle inventory.  
..           ...                                                ...  
734            Y    Matched an Active vehicle in vehicle inventory.  
733            Y    Matched an Active vehicle in vehicle inventory.  
732            Y    Matched an Active vehicle in vehicle inventory.  
735            Y    Matched an Active vehicle in vehicle inventory.  
740            Y    Matched an Active vehicle in vehicle inventory.  

[741 rows x 4 columns]

*OPTION 2: THE OUTPUT IS A LIST OF ONLY THE VINS THAT DON'T MATCH*

In [12]:
inv_agencies = rev_veh_inventory['Organization'].unique()
a30_agencies = a30['Organization'].unique()

output = []
for agency in a30_agencies:
    if len(a30[a30['Organization']==agency]) > 0:
        vins_a30 = a30[a30['Organization']==agency]['VIN'].unique() #will list only active ins
        vins_inventory = rev_veh_inventory[(rev_veh_inventory['Organization']==agency) \
                 & (rev_veh_inventory['Status']=='Active')]['VIN'].unique()
        
        for v in vins_a30: #check whether each VIN exists in inventory
            if v not in vins_inventory:
                description = f"Not an active vehicle in this org's inventory. Investigate."
                result = "N"
                
                output_line = {"Organization": agency,
                           "VIN": v,
                           "check_status": result,
                          "Description": description}
                output.append(output_line)
            else:
                pass
    
    mismatched_vin_checklist = pd.DataFrame(output).sort_values(by="Organization")


In [14]:
## TO DO: add a check for whether it's in the inventory at all with non-active status - if grants team wants it.

print(mismatched_vin_checklist.shape)
mismatched_vin_checklist.head()

(43, 4)


Organization                VIN check_status  \
0   Alpine County Community Development  1GB3G2BL5E1199290            N   
1                        Amador Transit  4UZADRDU4ECFR9869            N   
2                   City of Chowchilla   1FDFE4FS4CDA11717            N   
11                      City of Shafter  3C6TRVPGXFE512763            N   
10                      City of Shafter  3C6TRVPG8FE512762            N   

                                          Description  
0   Not an active vehicle in this org's inventory....  
1   Not an active vehicle in this org's inventory....  
2   Not an active vehicle in this org's inventory....  
11  Not an active vehicle in this org's inventory....  
10  Not an active vehicle in this org's inventory....

---
### Checks 2 & 3: reported totals of active vehicles in both, AND VOMS total from RR-20.  
Output can be that the totals match, or if not, list what each total is.  
  
```Org | n_a30_vehicles | n_rr20_VOMS | n_inventory_vehicles | check_result | notes``` 

In [15]:
# Import relevant part of the rr-20
rr20_service = pd.read_excel("data/NTD_Annual_Report_Rural_2022.xlsx", 
                     sheet_name="Service Data", index_col=None) 


In [16]:
print(rr20_service.shape)
rr20_service.head(5)

(92, 9)


Organization Legal Name Common Name/Acronym/DBA  Fiscal Year  \
0  Alpine County Community Development                     NaN         2022   
1                       Amador Transit                     NaN         2022   
2                       Amador Transit                     NaN         2022   
3                       Amador Transit                     NaN         2022   
4            Calaveras Transit Agency                      CTA         2022   

                               Mode  Annual VRM  Annual VRH  Annual UPT  \
0       Demand Response (DR) - (DO)     10386.0       643.0       384.0   
1          Commuter Bus (CB) - (DO)     45472.0      1637.0      1629.0   
2       Demand Response (DR) - (DO)     31337.0      2991.0      7028.0   
3  Deviated Fixed Route (DF) - (DO)    153757.0      7302.0     16196.0   
4       Demand Response (DR) - (PT)     23812.0      1416.0      2104.0   

   Sponsored UPT  VOMX  
0          274.0   1.0  
1            0.0   1.0  
2          315.0   5.0  
3            0.0  11.0  
4            0.0   5.0

In [17]:
agency = "Amador Transit"
rr20_service[rr20_service['Organization Legal Name']==agency]['VOMX'].sum()

17.0

In [18]:
inv_agencies = rev_veh_inventory['Organization'].unique()
a30_agencies = a30['Organization'].unique()
rr20_agencies = rr20_service['Organization Legal Name'].unique()

output = []
for agency in a30_agencies:
    if len(a30[a30['Organization']==agency]) > 0:
        a30_n = a30[a30['Organization']==agency]['VIN'].nunique()
        
    if len(rev_veh_inventory[rev_veh_inventory['Organization'] == agency]) > 0:
        inv_n = rev_veh_inventory[(rev_veh_inventory['Organization'] == agency) \
         & (rev_veh_inventory['Status']=='Active')]['VIN'].nunique()
    
    if len(rr20_service[rr20_service['Organization Legal Name']==agency]) > 0:
        rr20_n = rr20_service[rr20_service['Organization Legal Name']==agency]['VOMX'].sum()
        rr20_n = round(rr20_n)

    if (a30_n <= inv_n) & (rr20_n <= inv_n) & (a30_n >= rr20_n):
        result = "pass"
        description = "" #"VOMS & A-30 vehicles reported are equal to and/or lower than active inventory."
    elif (a30_n > inv_n):
        result = "warning"
        description = "More A-30 vehicles reported than in active inventory."
    elif (a30_n < rr20_n):
        result = "fail"
        description = "Total VOMS is greater than total A-30 vehicles reported."

    output_line = {"Organization": agency,
                   "n_a30_vehicles": a30_n,
                   "n_rr20_VOMS": rr20_n,
                   "n_inventory_vehicles": inv_n,
                   "check_result": result,
                   "Description": description}
    output.append(output_line)
    totals_checklist = pd.DataFrame(output)


In [19]:
print(totals_checklist.shape)
totals_checklist

(48, 6)


Organization  n_a30_vehicles  \
0                 Alpine County Community Development               2   
1                                      Amador Transit              17   
2                           Calaveras Transit Agency               10   
3                                      City of Arcata               6   
4                                       City of Arvin               6   
5                                      City of Auburn               7   
6                                 City of Chowchilla                3   
7            City of Corcoran - Corcoran Area Transit               6   
8                                       City of Dixon              12   
9                                    City of Escalon                3   
10                                  City of Guadalupe               5   
11                                  City of McFarland               2   
12                                    City of Needles               3   
13                                       City of Ojai               5   
14                                 City of Ridgecrest               9   
15                                  City of Rio Vista               5   
16                                    City of Shafter              11   
17                                    City of Solvang               6   
18                                       City of Taft               8   
19                                      City of Wasco               3   
20                       Colusa County Transit Agency              11   
21                                 County of Mariposa              11   
22  County of Nevada Public Works, Transit Service...              12   
23  County of Sacramento Department of Transportation              19   
24        County of Shasta Department of Public Works               2   
25                                 County of Siskiyou              10   
26                                   County of Tulare              22   
27                   Eastern Sierra Transit Authority              64   
28                 Fresno County Rural Transit Agency             115   
29             Glenn County Transportation Commission               8   
30                         Humboldt Transit Authority              31   
31             Kern Regional Transit (County of Kern)              60   
32                      Lassen Transit Service Agency              12   
33                                      Madera County              13   
34                        Mendocino Transit Authority              38   
35                        Modoc Transportation Agency               5   
36                    Morongo Basin Transit Authority              23   
37           Mountain Area Regional Transit Authority              29   
38                   Palo Verde Valley Transit Agency              10   
39            Plumas County Transportation Commission               9   
40                    Redwood Coast Transit Authority              13   
41   San Benito County Local Transportation Authority              28   
42            Sierra County Transportation Commission               4   
43                       Tehama County Transit Agency              17   
44                                    Town of Truckee               7   
45       Trinity County Department of Transportation                8   
46              Tuolumne County Transit Agency (TCTA)              22   
47      Yosemite Area Regional Transportation System                9   

    n_rr20_VOMS  n_inventory_vehicles check_result  \
0             1                     1         fail   
1            17                    19         pass   
2            11                    10         fail   
3             2                     6         pass   
4             4                     7         pass   
5             6                     7         pass   
6             2                     2         fail   
7             6                

---
### Export all draft outputs to excel files
We also add a few more columns for Liaisions to manually track agency responses.  
Again, plan is to show the grants and 5311 liaision team these output options to get their feedback.

In [29]:
# create a excel writer object
with pd.ExcelWriter("reports/voms_check_report.xlsx") as writer:
   
    full_vin_checklist.to_excel(writer, sheet_name="vin_check_full", index=False, startrow=2)
    mismatched_vin_checklist.to_excel(writer, sheet_name="vin_check_fails_only", index=False, startrow=2)
    totals_checklist.to_excel(writer, sheet_name="totals_check", index=False, startrow=2)
    
    workbook = writer.book
    worksheet1 = writer.sheets["vin_check_full"]
    worksheet2 = writer.sheets["vin_check_fails_only"]
    worksheet3 = writer.sheets["totals_check"]
    cell_highlight = workbook.add_format({
        'fg_color': 'yellow',
        'bold': True,
        'border': 1
    })
    
    report_title = "NTD Data Validation Report"
    title_format = workbook.add_format({
              'bold': True,
              'valign': 'center',
              'align': 'left',
              'font_color': '#1c639e',
              'font_size': 15
            })
    subtitle1 = "VOMS Inventory Vehicle Check: Validation Warnings"
    subtitle2 = "VOMS RR-20 & A-30 check"
    subtitle_format = workbook.add_format({
          'bold': True,
          'align': 'left',
          'font_color': 'black',
          'font_size': 19
        })

    # Column width and format - sheet 1
    worksheet1.set_column(0, 0, 35) #col A width
    worksheet1.set_column(1, 2, 20) #cols B-C width
    worksheet1.set_column(3, 3, 53) #col D width
    worksheet1.write('A1', report_title, title_format)
    worksheet1.write('A2', subtitle1, subtitle_format)
    worksheet1.freeze_panes('B4')
    
    # Column width and format - sheet 2
    worksheet2.set_column(0, 0, 35) #col A width
    worksheet2.set_column(1, 2, 20) #cols B-C width
    worksheet2.set_column(3, 5, 53) #cols D-F width
    worksheet2.write('E3', 'Agency Response', cell_highlight)
    worksheet2.write('F3', 'Response Date', cell_highlight)
    worksheet2.write('A1', report_title, title_format)
    worksheet2.write('A2', subtitle1, subtitle_format)
    worksheet2.freeze_panes('B4')

    
    # Column width and format - sheet 3
    worksheet3.set_column(0, 0, 35) #col A width
    worksheet3.set_column(1, 4, 18) #cols B-E width
    worksheet3.set_column(5, 7, 53) #cols F-H width
    worksheet3.write('G3', 'Agency Response', cell_highlight)
    worksheet3.write('H3', 'Response Date', cell_highlight)
    worksheet3.write('A1', report_title, title_format)
    worksheet3.write('A2', subtitle2, subtitle_format)
    worksheet3.freeze_panes('B4')